src/h2_streams.erl

```diff
@@ -0,0 +1,541 @@
+    %% can. Then, based on which number was the limiting factor, we'll
+    %% make another decision
+
+    %% If it was MAX_FRAME_SIZE, then we recurse into this same
+    %% function, because we're able to send another frame of some
+    %% length.
+
+    %% If it was connection send window size, we're blocked at the
+    %% connection level and we should break out of this recursion
+
+    %% If it was stream send_window size, we're blocked on this
+    %% stream, but other streams can still go, so we'll break out of
+    %% this recursion, but not the connection level
+
+    SSWS = Stream#active_stream.send_window_size,
+
+    QueueSize = byte_size(Stream#active_stream.queued_data),
+
+    {MaxToSend, ExitStrategy} =
+        case {MFS =< SWS andalso MFS =< SSWS, SWS < SSWS} of
+            %% If MAX_FRAME_SIZE is the smallest, send one and recurse
+            {true, _} ->
+                {MFS, max_frame_size};
+            {false, true} ->
+                {SWS, connection};
+            _ ->
+                {SSWS, stream}
+        end,
+
+    {Frame, SentBytes, NewS} =
+        case MaxToSend > QueueSize of
+            true ->
+                Flags = case Stream#active_stream.body_complete of
+                         true -> ?FLAG_END_STREAM;
+                         false -> 0
+                        end,
+                %% We have the power to send everything
+                {{#frame_header{
+                     stream_id=Stream#active_stream.id,
+                     flags=Flags,
+                     type=?DATA,
+                     length=QueueSize
+                    },
+                  http2_frame_data:new(Stream#active_stream.queued_data)}, %% Full Body
+                 QueueSize,
+                 Stream#active_stream{
+                   queued_data=done,
+                   send_window_size=SSWS-QueueSize}};
+            false ->
+                <<BinToSend:MaxToSend/binary,Rest/binary>> = Stream#active_stream.queued_data,
+                {{#frame_header{
+                     stream_id=Stream#active_stream.id,
+                     type=?DATA,
+                     length=MaxToSend
+                    },
+                  http2_frame_data:new(BinToSend)},
+                 MaxToSend,
+                 Stream#active_stream{
+                   queued_data=Rest,
+                   send_window_size=SSWS-MaxToSend}}
+        end,
+
+    _Sent = http2_stream:send_frame(Stream#active_stream.pid, Frame),
+
+    case ExitStrategy of
+        max_frame_size ->
+            s_send_what_we_can(SWS - SentBytes, MFS, NewS);
+        stream ->
+            {SWS - SentBytes, NewS};
+        connection ->
+            {0, NewS}
+    end;
+s_send_what_we_can(SWS, _MFS, NonActiveStream) ->
+    {SWS, NonActiveStream}.
+
+-ifdef(TEST).
+-compile([export_all]).
+basic_streams_structure_test() ->
+    %% Constructor:
+    _Streams = new(client), %% There will be more to this
+
```

Comment:
FYI, this test is just some logical flow for memory managed version, and should be uncommented as part of that work